### PR TITLE
Blended query optimization

### PR DESCRIPTION
**Pull request by gl3nn**
There's still an issue with sets that I'm trying to solve but it would be nice to get a review already.

### CODE REVIEW

File: fireant/queries/builder/dataset_blender_query_builder.py [hunk 1]
```diff
@@ -204,6 +235,9 @@ def _map_fields(fields):
         orders=[],
         share_dimensions=dataset_share_dimensions,
     )
+    if result[0].get_sql() == "":
+        return []
```
Comment (mikeengland) [inline fireant/queries/builder/dataset_blender_query_builder.py:239]: Out of interest, what is the case where the sql can be an empty string?

File: fireant/queries/builder/dataset_blender_query_builder.py [hunk 2]
```diff
@@ -343,35 +385,133 @@ def sql(self):
...
+            selected_blender_dimensions, operations
+        )
+        non_set_filters = omit_set_filters(self._filters)
+
+        # Determine for each dataset which metrics and dimensions need to be selected
+        dataset_dimensions = [[] for _ in range(len(datasets))]
+        dataset_metrics = []
+        dataset_filters = []
+        dataset_included_in_final_query = [False] * len(datasets)
+
```
Comment (mikeengland) [inline fireant/queries/builder/dataset_blender_query_builder.py:420]: Thanks for these step by step comments, these make it much easier to understand!

**Comment by mikeengland** [comment]
The changes look good to me so far - it was very easy to read through the logic.

File: fireant/queries/builder/dataset_blender_query_builder.py (hunk 1 above)
Comment (gl3nn) [reply fireant/queries/builder/dataset_blender_query_builder.py]: I had some cases before but I think with all the later optimisation commits this should no longer happen. I removed it from the code so we can test whether it occurs.

**Review by mikeengland — APPROVED** [review]

**Pull request merged**